plugin/evm/vm.go

```diff
@@ -620,8 +621,21 @@ func (vm *VM) Shutdown(context.Context) error {
 }
 
 // buildBlock builds a block to be wrapped by ChainState
-func (vm *VM) buildBlock(context.Context) (snowman.Block, error) {
-	block, err := vm.miner.GenerateBlock()
+func (vm *VM) buildBlock(ctx context.Context) (snowman.Block, error) {
+	return vm.buildBlockWithContext(ctx, nil)
+}
+
+func (vm *VM) buildBlockWithContext(ctx context.Context, proposerVMBlockCtx *block.Context) (snowman.Block, error) {
+	if proposerVMBlockCtx != nil {
+		log.Debug("Building block with context", "pChainBlockHeight", proposerVMBlockCtx.PChainHeight)
+	} else {
+		log.Debug("Building block without context")
+	}
+
+	// Verify any transaction predicates with the given block context.
+	pendingTxs := vm.txPool.PendingWithPredicates(true, vm.currentRules(), vm.ctx, proposerVMBlockCtx)
```

Comment:
Using `vm.currentRules()` will use the rules for the current head block of the chain, which will in most cases be different from the block that is about to be built. Since the block being built may have a different timestamp, a different network upgrade could have activated and the ruleset could change.

As a result, we should make sure to calculate the correct `params.Rules` for the actual block we are building or else we may apply an incorrect rule set here, which could halt the chain.

We may need to revert part of the change here and call `PendingWithPredicates` inside of `miner.GenerateBlock` and pass in the P-Chain context instead since that is where the newly created block is created and assigned a block number and timestamp.

Reply:
I modified `miner.GenerateBlock` to take the `snowCtx` and `proposerVMBlockCtx`, which are eventually passed to the predicates before the block is built. We also use the new block's rules when checking the predicate.